### ISSUE TITLE
404 error, not finding html during webcomponent unit test

### ISSUE THREAD

**Issue by jimisdrpc** [open-wc/open-wc#730]
Disclaimer: posted Yesterday in https://stackoverflow.com/questions/57468253/404-error-not-finding-html-during-webcomponent-unit-test. I am not sure if is an issue with OpenWC or a gape in my knowlodge how use Karma for testing javascript that imports a html file.

When I run karma testing a javascript file which imports a HTML file I expected the html available for testing. It is not clear for me if I am trying something that it is not karma prepared for or I should somehow add the HTML file before estarting the test.

Put in few words, what I should do in order to unit test bellow webcomponent which is splited in two files (js and html)? Maybe an usefull comment will be if I misunderstood something wrong about karma (for instance, it can't load a file imported by javascript assyncronously)

Console error:

`# npm run test

> skyscanner-openwc-graphql@0.0.0 test C:\_d\WSs\rapidapi-vanilla-webcomponents\skyscanner-openwc-graphql
> karma start --coverage

START:
12 08 2019 15:29:40.982:WARN [filelist]: Pattern "C:/_d/WSs/rapidapi-vanilla-webcomponents/skyscanner-openwc-graphql/__snapshots__/**/*.md" does not match any file.
12 08 2019 15:29:41.005:INFO [karma-server]: Karma v4.2.0 server started at http://0.0.0.0:9876/
12 08 2019 15:29:41.007:INFO [launcher]: Launching browsers ChromeHeadlessNoSandbox with concurrency unlimited
12 08 2019 15:29:41.013:INFO [launcher]: Starting browser ChromeHeadless
12 08 2019 15:29:42.172:INFO [HeadlessChrome 76.0.3809 (Windows 10.0.0)]: Connected on socket fPyJTkm8LvWTrPBoAAAA with id 27628633
12 08 2019 15:29:47.926:WARN [web-server]: 404: /src/skyscanner-flight-search/skyscanner-flight-search.html
  skyscanner flight search
    × show div
HeadlessChrome 76.0.3809 (Windows 10.0.0) skyscanner flight search show div FAILED
        TypeError: Cannot read property 'querySelector' of null
            at Context.querySelector (test/skyscanner-flight-search.test.js:10:23)

Finished in 0.102 secs / 0.022 secs @ 15:29:48 GMT-0300 (GMT-03:00)

SUMMARY:
√ 0 tests completed
× 1 test failed

FAILED TESTS:
  skyscanner flight search
    × show div
      HeadlessChrome 76.0.3809 (Windows 10.0.0)
    TypeError: Cannot read property 'querySelector' of null
        at Context.querySelector (test/skyscanner-flight-search.test.js:10:23)
`

Webcomponent, javascript file (skyscanner-flight-search.js):

`(async() => {
    const res = await fetch(
        "../src/skyscanner-flight-search/skyscanner-flight-search.html"
    );
    const textTemplate = await res.text();

    const HTMLTemplate = new DOMParser()
        .parseFromString(textTemplate, "text/html")
        .querySelector("template");

    class skyscannerFlightSearch extends HTMLElement {
        constructor() {
            super();
        }

        connectedCallback() {
            this.attachShadow({ mode: "open" });
            this.shadowRoot.appendChild(HTMLTemplate.content.cloneNode(true));

            const inputSessionKey = this.shadowRoot.getElementById("inputSessionKey");
            //const url = "http://localhost:8080/getsessionkey";
            const url = "https://rapid-api-to-learn.herokuapp.com/getsessionkey";

            fetch(url)
                .then(function(body) {
                    return body.text();
                })
                .then(function(data) {
                    inputSessionKey.value = data;
                    console.log(data);
                });
        }
    }

    window.customElements.define(
        "skyscanner-flight-search",
        skyscannerFlightSearch
    );
})();
`

Webcomponent, html file (skyscanner-flight-search.html)

`<template id="skyscanner-flight-search-template"
  ><div id="firstdiv"><input id="inputSessionKey" /></div
></template>
`

Dependencies (just the relevant ones)

`    "scripts": {
...
        "test": "karma start --coverage",
...

    },
    "dependencies": {
        "lit-html": "^1.0.0",
        "lit-element": "^2.0.1"
    },
    "devDependencies": {
        "es-dev-server": "^1.5.0",
        "@open-wc/eslint-config": "^1.0.0",
        "@open-wc/prettier-config": "^0.1.10",
        "husky": "^1.0.0",
        "lint-staged": "^8.0.0",
        "@open-wc/testing-karma": "^3.0.0",
        "webpack-merge": "^4.1.5",
        "@open-wc/testing-karma-bs": "^1.0.0",
        "@open-wc/testing": "^0.11.1",
        "@open-wc/demoing-storybook": "^0.3.0"
    },
`

karma.conf.js

`const { createDefaultConfig } = require('@open-wc/testing-karma');
const merge = require('webpack-merge');

module.exports = config => {
    config.set(
        merge(createDefaultConfig(config), {
            files: [

                { pattern: config.grep ? config.grep : 'test/**/*-search.test.js', type: 'module' },
            ],

            esm: {
                nodeResolve: true,
            },
            // you can overwrite/extend the config further
        }),
    );
    return config;
};`

Apologize for bad formatting. I don't know why adding inside <> doesn't format properly

**Comment by LarsDenBakker** [comment]
You can format blocks of code by surrounding it with three ` backticks.

The problem is that you're fetching a file using an absolute path, this assumption does not always hold true depending on where the code is served from and after bundling.

My suggestion is to use `import.meta.url`, to resolve the url relative to the module:
```js
const htmlPath = new URL("../src/skyscanner-flight-search/skyscanner-flight-search.html", import.meta.url);
const res = await fetch(htmlPath);
```

**Comment by jimisdrpc** [comment]
I tried your suggestion but in fafct now I can't either run neither test the webcomponent. With my above code I was able to run at least and the exception I only got during tests.

Kindly, see the error with your suggestion even when just "npm start".


![import meta](https://user-images.githubusercontent.com/16684247/62980890-30d32e80-bdfe-11e9-8e46-c55271ed6ff0.gif)

**Comment by LarsDenBakker** [comment]
This is probably because `../src/skyscanner-flight-search/skyscanner-flight-search.html` isn't pointing to your HTML files relative to your module. Sorry I should have made that clearer in the example.

Write out the path how to get to your html file relative to your module. If they're in the same folder, it should be:

```js
const htmlPath = new URL("./skyscanner-flight-search.html", import.meta.url);
```

**Comment by jimisdrpc** [comment]
Thanks. Now I have exactly same issue mentioned originally in my question. I mean I can run the webcomponent but I can't unit test it. I get the message:

'FAILED TESTS:
  skyscanner flight search
    × show div
      HeadlessChrome 76.0.3809 (Windows 10.0.0)
    TypeError: Cannot read property 'querySelector' of null
        at Context.querySelector (test/skyscanner-flight-search.test.js:10:23)'

No matter if my javascript weebcomponent fetch with or without import.meta.url I will get the error pasted above. I guess because textTemplate variable ins't rerally receiving the text from my HTML file.

Have you coded and successfully unit test a webcomponent separating html from javascript somehow? Would you mind share a Hello World with me? Please, note that I can make the webcomponent run (npm start) but I can't test it (npm run test) only when I split the webcomponent in two files (js and html). Maybe there is some base concept regard unit test javascript which depends on another file that I am missing when using Karma. Are you aware about that?

**Comment by LarsDenBakker** [comment]
Yes we have a test using `import.meta.url`. The code: 
https://github.com/open-wc/open-wc/blob/master/packages/karma-esm/demo/basic/src/render-image.js

and a test: https://github.com/open-wc/open-wc/blob/master/packages/karma-esm/demo/basic/test/render-image.test.js which doesn't give a 404 when it runs.

Are you sure you're using a relative path, not an absolute path for the file?

**Comment by jimisdrpc** [comment]
I pushed whole project in https://github.com/jimisdrpc/skyscanner-openwc-graphql.git. Basically it is a scaffolded project where I added two Vanilla Webcomponents in src\skyscanner-flight-search folder: (1) my-complete-webcomponent contains html and javascript in same file and (2) skyscanner-flight-search splitted in html and javascript files. I can unit test the one containing both javascript and html in same file but other fails with error mentioned above.

**Comment by jimisdrpc** [comment]
Sorry, maybe you didn't understand me beucase I didn't write clearly. The example bellow isn't a sample of a WebComponent splitted in two files being one javascript and another html. Take a note that you are creating the element and then just addint to it an image from another file. As far as I can see, complete different idea of putting all your static text in one file (HTML file) and then fetch it from a javascript.

 export function renderImage() {
  const dir = new URL('./', import.meta.url);
  **const img = document.createElement('img');
  img.src = `${dir}logo.png`;**
  document.body.appendChild(img);
}

**Comment by LarsDenBakker** [comment]
The concept is fetching a file relative to your web cmoponent, what you do with the file's content should not matter.

I checked out the project locally, it's working for me with:
```js
    const htmlPath = new URL(
        './skyscanner-flight-search.html',
        import.meta.url,
    );
    const res = await fetch(htmlPath);
```

You can see you get a HTTP 200 response both with the dev server and in karma.

However I see that there is a race condition. By the time you're running your test assertion you can't be certain the sync work inside your component is done yet. So sometimes it hasn't fetched the template yet.

You can wait for your custom elements to be defined before running your test:

```js
await customElements.whenDefined('skyscanner-flight-search');
```

**Comment by LarsDenBakker** [comment]
https://github.com/tc39/proposal-top-level-await will help a lot with this scenario

**Comment by jimisdrpc** [comment]
I will study this page and return soon my findings. Meanwhile, with two different files approach in mind. what is your opinion about I SUCCESSFULLY run the webcomponent but I FAILED to unit test? Am I trying something Karma isn't the proper tool for such scenario? I mean, it seems it is a problem with assyncronously trying to import text from a file and when unit testing it the text isn't available yet. I read somewhere someone saying to put a delay but it is ridiculius delaying if this is going to delay also in production and it doesn't seem to be a problem when running the webcomponent, only when testing. Maybe I should add some exteernal framework like Webpack to package both files in single unit test. Any idea will be wellcomed

**Comment by LarsDenBakker** [comment]
I think it's fine, you just need some way in your test to know when your component finished loading. Using `await customElements.whenDefined('skyscanner-flight-search');` is fine for this.

**Comment by LarsDenBakker** [comment]
For production optimizations you will want to look into some way of bundling the files, though.

**Comment by jimisdrpc** [comment]
BINGO! You fixed my issue! Thanks! I have to do two things in order to unit test my vanilla webcomponent spplited in two files:

1) my webcomponent javascript 
'
    const htmlPath = new URL(
        './skyscanner-flight-search.html',
        import.meta.url,
    );
    const res = await fetch(htmlPath);
...
'

2) my unit test

'
import { html, fixture, expect } from '@open-wc/testing';

import '../src/skyscanner-flight-search/skyscanner-flight-search.js';

describe('skyscanner flight search', () => {
    it('show div', async() => {
        const el = await fixture(html `
      <skyscanner-flight-search></skyscanner-flight-search>
    `);

        window.customElements.whenDefined('skyscanner-flight-search').then(() => {
            el.shadowRoot.querySelector('#firstdiv');
            expect(el).to.exist;
        })

    });
});
'
No I am facing issue with Istambul Coverage perentage but this is a topic for another issue after some google search.

From cmd console:

'
...
14 08 2019 12:16:36.156:INFO [HeadlessChrome 76.0.3809 (Windows 10.0.0)]: Connected on socket QLSxFM6Utha7Y-eDAAAA with id 62450035
  skyscanner flight search
    √ show div

Finished in 0.033 secs / 0.004 secs @ 12:16:40 GMT-0300 (GMT-03:00)

SUMMARY:
√ 1 test completed
TOTAL: 1 SUCCESS

=============================== Coverage summary ===============================
Statements   : 26.67% ( 4/15 )
Branches     : 100% ( 0/0 )
Functions    : 20% ( 1/5 )
Lines        : 26.67% ( 4/15 )
================================================================================
14 08 2019 12:16:40.577:ERROR [reporter.coverage-istanbul]: Coverage for statements (26.67%) does not meet global threshold (80%)
14 08 2019 12:16:40.578:ERROR [reporter.coverage-istanbul]: Coverage for lines (26.67%) does not meet global threshold (80%)
14 08 2019 12:16:40.579:ERROR [reporter.coverage-istanbul]: Coverage for functions (20%) does not meet global threshold (80%)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! skyscanner-openwc-graphql@0.0.0 test: `karma start --coverage`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the skyscanner-openwc-graphql@0.0.0 test script.
...
'
From debug.log

'
14 08 2019 12:16:36.156:INFO [HeadlessChrome 76.0.3809 (Windows 10.0.0)]: Connected on socket QLSxFM6Utha7Y-eDAAAA with id 62450035
  skyscanner flight search
    √ show div

Finished in 0.033 secs / 0.004 secs @ 12:16:40 GMT-0300 (GMT-03:00)

SUMMARY:
√ 1 test completed
TOTAL: 1 SUCCESS

=============================== Coverage summary ===============================
Statements   : 26.67% ( 4/15 )
Branches     : 100% ( 0/0 )
Functions    : 20% ( 1/5 )
Lines        : 26.67% ( 4/15 )
================================================================================
14 08 2019 12:16:40.577:ERROR [reporter.coverage-istanbul]: Coverage for statements (26.67%) does not meet global threshold (80%)
14 08 2019 12:16:40.578:ERROR [reporter.coverage-istanbul]: Coverage for lines (26.67%) does not meet global threshold (80%)
14 08 2019 12:16:40.579:ERROR [reporter.coverage-istanbul]: Coverage for functions (20%) does not meet global threshold (80%)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! skyscanner-openwc-graphql@0.0.0 test: `karma start --coverage`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the skyscanner-openwc-graphql@0.0.0 test script.
'

13 verbose stack Error: skyscanner-openwc-graphql@0.0.0 test: `karma start --coverage`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:189:13)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:189:13)
13 verbose stack     at maybeClose (internal/child_process.js:970:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
14 verbose pkgid skyscanner-openwc-graphql@0.0.0

**Comment by jimisdrpc** [comment]
Sorry. I mentioned above that I have fixed my issue but I just realized that no matter which test I tried between the promisse window.customElements.whenDefined isn't really been triggered.

**Comment by jimisdrpc** [comment]
'describe('skyscanner flight search', () => {
    it('show div', async() => {
        const el = await fixture(html `
      <skyscanner-flight-search></skyscanner-flight-search>
    `);
        console.debug('before promisse');
        window.customElements.whenDefined('skyscanner-flight-search').then(() => {
            el.shadowRoot.querySelector('#firstdiv2');
            console.debug('after promisse');
            expect(el).to.exist;

        });
    });
});'

The message "After Promise" is never printed

**Comment by LarsDenBakker** [comment]
In mocha you need to return a promise if you want to wait on it. I suggest making it an async function and using await, since that always returns a promise

**Comment by jimisdrpc** [comment]
Sorry, it is not clear for me your suggestion. Actually, it is already wrapped in an Async function (it('show div', async() => {...)
